File: packages/simulations/typescript/reference-model.ts
import {
  auctionPriceAt,
  currentTotalSupply,
  earnedStrategyReward,
  miningRateAt,
  netSupplyChange,
  nextAuctionInitPrice,
  previewRedemption,
  projectTotalSupply,
  quoteMiningAccrual,
  quoteMiningPayment,
  quoteMiningPrice,
  redemptionPercentageWad,
  settleStrategyPayment,
  updateRewardIndex,
} from '@gumball-6900/sdk';

type StringRecord = Record<string, unknown>;
const INITIAL_SUPPLY = 0n;

export interface ReferenceScenarios extends StringRecord {
  schemaVersion: string;
  usdGDecimals: string;
  targetDecimals: string;
  miningCases: StringRecord[];
  auctionCases: StringRecord[];
  rewardCases: StringRecord[];
  redemptionCases: StringRecord[];
  supplyCases: StringRecord[];
}

const record = (value: unknown, label: string): StringRecord => {
  if (typeof value !== 'object' || value === null || Array.isArray(value))
    throw new TypeError(`${label} must be object`);
  return value as StringRecord;
};
const text = (value: unknown, label: string): string => {
  if (typeof value !== 'string' || !/^-?\d+$/u.test(value))
    throw new TypeError(`${label} must be a decimal integer string`);
  return value;
};
const id = (value: unknown, label: string): string => {
  if (typeof value !== 'string' || value.length === 0) throw new TypeError(`${label} must be a non-empty string`);
  return value;
};
const array = (value: unknown, label: string): unknown[] => {
  if (!Array.isArray(value)) throw new TypeError(`${label} must be array`);
  return value;
};

export function parseReferenceScenarios(value: unknown): ReferenceScenarios {
  const root = record(value, 'scenarios');
  const parseCases = (key: string): StringRecord[] =>
    array(root[key], key).map((entry, index) => {
      const parsed = record(entry, `${key}[${index}]`);
      id(parsed.id, `${key}[${index}].id`);
      return parsed;
    });
  return {
    schemaVersion: text(root.schemaVersion, 'schemaVersion'),
    usdGDecimals: text(root.usdGDecimals, 'usdGDecimals'),
    targetDecimals: text(root.targetDecimals, 'targetDecimals'),
    miningCases: parseCases('miningCases'),
    auctionCases: parseCases('auctionCases'),
    rewardCases: parseCases('rewardCases'),
    redemptionCases: parseCases('redemptionCases'),
    supplyCases: parseCases('supplyCases'),
  };
}

const big = (entry: StringRecord, key: string): bigint => BigInt(text(entry[key], key));
const decimal = (value: bigint): string => value.toString();

function synchronizedMiningEmission(
  elapsedSinceStart: bigint,
  curve: { halvingPeriod: bigint; initialTps: bigint; tailTps: bigint },
): bigint {
  if (elapsedSinceStart < 0n || curve.halvingPeriod <= 0n || curve.initialTps <= 0n || curve.tailTps <= 0n) {
    throw new RangeError('invalid synchronized mining input');
  }

  let emission = 0n;
  let remaining = elapsedSinceStart;
  for (let era = 0n; era < 256n; era += 1n) {
    if (remaining === 0n) return emission;
    const shifted = curve.initialTps >> era;
    if (shifted <= curve.tailTps) return emission + curve.tailTps * remaining;
    const activeSeconds = remaining < curve.halvingPeriod ? remaining : curve.halvingPeriod;
    emission += shifted * activeSeconds;
    remaining -= activeSeconds;
  }
  throw new RangeError('positive tail must be reached within uint256 shift bounds');
}

export function computeReferenceResults(scenarios: ReferenceScenarios) {
  const miningQuotes = scenarios.miningCases.map((entry) => {
    const slotTps = array(entry.slotTps, 'slotTps').map((value) => BigInt(text(value, 'slotTps')));
    const accrual = quoteMiningAccrual({ elapsedSeconds: big(entry, 'accrualSeconds'), slotTps });
    const curve = {
      halvingPeriod: big(entry, 'halvingPeriod'),
      initialTps: big(entry, 'initialTps'),
      tailTps: big(entry, 'tailTps'),
    };
    const nextGlobalTps = miningRateAt(big(entry, 'elapsedSinceStart'), curve);
    const synchronizedEmission = synchronizedMiningEmission(big(entry, 'elapsedSinceStart'), curve);
    const payment = quoteMiningPayment(big(entry, 'payment'), entry.hasPreviousMiner === true);
    return {
      id: id(entry.id, 'id'),
      price: decimal(quoteMiningPrice(big(entry, 'initialPrice'), big(entry, 'elapsedSeconds'))),
      previousMinerAmount: decimal(payment.previousMinerAmount),
      resonanceAmount: decimal(payment.resonanceAmount),
      slotEmissions: accrual.slotEmissions.map(decimal),
      totalEmission: decimal(accrual.totalEmission),
      nextGlobalTps: decimal(nextGlobalTps),
      nextSlotTps: decimal(nextGlobalTps / 16n),
      synchronizedMiningEmission: decimal(synchronizedEmission),
      synchronizedGrossSupply: decimal(INITIAL_SUPPLY + synchronizedEmission),
    };
  });

  const auctionQuotes = scenarios.auctionCases.map((entry) => {
    const payment = auctionPriceAt(big(entry, 'initPrice'), big(entry, 'elapsedSeconds'), big(entry, 'epochPeriod'));
    const bribeBasisPoints = big(entry, 'bribeBps');
    const partitionBribeBasisPoints = array(entry.paymentPartitionBps, 'paymentPartitionBps').map((raw) =>
      BigInt(text(raw, 'paymentPartitionBps')),
    );
    const paymentPartitions = array(entry.paymentPartitions, 'paymentPartitions');
    if (partitionBribeBasisPoints.length !== paymentPartitions.length) {
      throw new RangeError('every payment partition needs one Bribe rate');
    }
    const settlement = settleStrategyPayment(big(entry, 'actualTargetReceived'), bribeBasisPoints);
    let partitionFundAmount = 0n;
    let partitionBribeAmount = 0n;
    for (const [index, raw] of paymentPartitions.entries()) {
      const part = BigInt(text(raw, 'paymentPartitions'));
      const classified = settleStrategyPayment(part, partitionBribeBasisPoints[index]!);
      partitionFundAmount += classified.fundAmount;
      partitionBribeAmount += classified.bribeAmount;
    }
    return {
      id: id(entry.id, 'id'),
      paymentAmount: decimal(payment),
      nextInitPrice: decimal(nextAuctionInitPrice(payment, big(entry, 'priceMultiplier'), big(entry, 'minInitPrice'))),
      bribeBasisPoints: decimal(bribeBasisPoints),
      fundAmount: decimal(settlement.fundAmount),
      bribeAmount: decimal(settlement.bribeAmount),
      partitionFundAmount: decimal(partitionFundAmount),
      partitionBribeAmount: decimal(partitionBribeAmount),
      partitionBribeBasisPoints: partitionBribeBasisPoints.map(decimal),
    };
  });

  const rewardQuotes = scenarios.rewardCases.map((entry) => {
    const update = updateRewardIndex(
      big(entry, 'rewardAmount'),
      big(entry, 'totalActiveWeight'),
      big(entry, 'precision'),
    );
    return {
      id: id(entry.id, 'id'),
      rewardPerWeightIncrement: decimal(update.rewardPerWeightIncrement),
      indexedReward: decimal(update.indexedReward),
      residue: decimal(update.residue),
      userEarned: decimal(
        earnedStrategyReward(
          big(entry, 'userActiveWeight'),
          update.rewardPerWeightIncrement,
          big(entry, 'userRewardPerWeightPaid'),
          big(entry, 'userAccrued'),
          big(entry, 'precision'),
        ),
      ),
    };
  });

  const redemptionQuotes = scenarios.redemptionCases.map((entry) => {
    const shares = big(entry, 'shares');
    const supplyBefore = big(entry, 'supplyBefore');
    const assets = array(entry.assets, 'assets').map((asset) => {
      const parsed = record(asset, 'asset');
      return { asset: id(parsed.asset, 'asset'), balance: BigInt(text(parsed.balance, 'balance')) };
    });
    return {
      id: id(entry.id, 'id'),
      percentageWad: decimal(redemptionPercentageWad(shares, supplyBefore)),
      outputs: previewRedemption(shares, supplyBefore, assets).map(({ asset, amount }) => ({
        asset,
        amount: decimal(amount),
      })),
    };
  });

  const supplyQuotes = scenarios.supplyCases.map((entry) => {
    const lifetimeMinted = big(entry, 'lifetimeMinted');
    const lifetimeBurned = big(entry, 'lifetimeBurned');
    const supply = currentTotalSupply(lifetimeMinted, lifetimeBurned);
    return {
      id: id(entry.id, 'id'),
      currentSupply: decimal(supply),
      netSupplyChange: decimal(netSupplyChange(big(entry, 'gbxMined'), big(entry, 'gbxBurned'))),
      projectedSupply: decimal(projectTotalSupply(supply, big(entry, 'gbxMined'), big(entry, 'gbxBurned'))),
    };
  });

  return {
    schemaVersion: scenarios.schemaVersion,
    usdGDecimals: scenarios.usdGDecimals,
    targetDecimals: scenarios.targetDecimals,
    infiniteSupply: true,
    initialSupply: INITIAL_SUPPLY.toString(),
    miningQuotes,
    auctionQuotes,
    rewardQuotes,
    redemptionQuotes,
    supplyQuotes,
  };
}

File: packages/simulations/python/reference_model.py
"""Independent reference calculations for committed Mine scenarios."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

WAD = 10**18
GENESIS_LP_GBX = 20_000_000 * WAD


def mul_div(a: int, b: int, denominator: int) -> int:
    return a * b // denominator


def mining_price(initial: int, elapsed: int) -> int:
    return 0 if elapsed >= 3_600 else initial - mul_div(initial, elapsed, 3_600)


def mining_rate(elapsed_since_start: int, initial: int, halving_period: int, tail: int) -> int:
    halvings = elapsed_since_start // halving_period
    return max(initial >> halvings, tail)


def synchronized_mining_emission(
    elapsed_since_start: int, initial: int, halving_period: int, tail: int
) -> int:
    if elapsed_since_start < 0 or halving_period <= 0 or initial <= 0 or tail <= 0:
        raise ValueError("invalid synchronized mining input")

    emission = 0
    remaining = elapsed_since_start
    for era in range(256):
        if remaining == 0:
            return emission
        shifted = initial >> era
        if shifted <= tail:
            return emission + tail * remaining
        active_seconds = min(remaining, halving_period)
        emission += shifted * active_seconds
        remaining -= active_seconds
    raise ValueError("positive tail must be reached within uint256 shift bounds")


def auction_price(initial: int, elapsed: int, duration: int) -> int:
    return 0 if elapsed >= duration else initial - mul_div(initial, elapsed, duration)


def classify_payment(payment: int, remainder: int = 0, bribe_bps: int = 1_000) -> tuple[int, int, int]:
    if payment < 0 or remainder < 0 or remainder >= 10_000 or bribe_bps < 0 or bribe_bps > 2_000:
        raise ValueError("invalid payment classification input")
    base_bribe, raw_remainder = divmod(payment * bribe_bps, 10_000)
    bribe_carry, next_remainder = divmod(remainder + raw_remainder, 10_000)
    bribe = base_bribe + bribe_carry
    return payment - bribe, bribe, next_remainder


def compute(scenarios: dict[str, Any]) -> dict[str, Any]:
    mining_quotes = []
    for case in scenarios["miningCases"]:
        payment = int(case["payment"])
        previous = mul_div(payment, 8_000, 10_000) if case["hasPreviousMiner"] else 0
        seconds = int(case["accrualSeconds"])
        emissions = [int(rate) * seconds for rate in case["slotTps"]]
        next_global = mining_rate(
            int(case["elapsedSinceStart"]),
            int(case["initialTps"]),
            int(case["halvingPeriod"]),
            int(case["tailTps"]),
        )
        synchronized_emission = synchronized_mining_emission(
            int(case["elapsedSinceStart"]),
            int(case["initialTps"]),
            int(case["halvingPeriod"]),
            int(case["tailTps"]),
        )
        mining_quotes.append(
            {
                "id": case["id"],
                "price": str(mining_price(int(case["initialPrice"]), int(case["elapsedSeconds"]))),
                "previousMinerAmount": str(previous),
                "resonanceAmount": str(payment - previous),
                "slotEmissions": [str(value) for value in emissions],
                "totalEmission": str(sum(emissions)),
                "nextGlobalTps": str(next_global),
                "nextSlotTps": str(next_global // 16),
                "synchronizedMiningEmission": str(synchronized_emission),
                "synchronizedGrossSupply": str(GENESIS_LP_GBX + synchronized_emission),
            }
        )

    auction_quotes = []
    for case in scenarios["auctionCases"]:
        payment = auction_price(int(case["initPrice"]), int(case["elapsedSeconds"]), int(case["epochPeriod"]))
        next_price = max(mul_div(payment, int(case["priceMultiplier"]), WAD), int(case["minInitPrice"]))
        bribe_bps = int(case["bribeBps"])
        partition_bps = [int(value) for value in case["paymentPartitionBps"]]
        if len(partition_bps) != len(case["paymentPartitions"]):
            raise ValueError("every payment partition needs one Bribe rate")
        fund, bribe, remainder = classify_payment(int(case["actualTargetReceived"]), 0, bribe_bps)
        partition_fund = 0
        partition_bribe = 0
        partition_remainder = 0
        for part, part_bps in zip(case["paymentPartitions"], partition_bps, strict=True):
            part_fund, part_bribe, partition_remainder = classify_payment(
                int(part), partition_remainder, part_bps
            )
            partition_fund += part_fund
            partition_bribe += part_bribe
        auction_quotes.append(
            {
                "id": case["id"],
                "paymentAmount": str(payment),
                "nextInitPrice": str(next_price),
                "bribeBasisPoints": str(bribe_bps),
                "fundAmount": str(fund),
                "bribeAmount": str(bribe),
                "splitRemainder": str(remainder),
                "partitionFundAmount": str(partition_fund),
                "partitionBribeAmount": str(partition_bribe),
                "partitionRemainder": str(partition_remainder),
                "partitionBribeBasisPoints": [str(value) for value in partition_bps],
            }
        )

    reward_quotes = []
    for case in scenarios["rewardCases"]:
        reward = int(case["rewardAmount"])
        weight = int(case["totalActiveWeight"])
        precision = int(case["precision"])
        increment = mul_div(reward, precision, weight)
        indexed = mul_div(increment, weight, precision)
        earned = int(case["userAccrued"]) + mul_div(
            int(case["userActiveWeight"]), increment - int(case["userRewardPerWeightPaid"]), precision
        )
        reward_quotes.append(
            {
                "id": case["id"],
                "rewardPerWeightIncrement": str(increment),
                "indexedReward": str(indexed),
                "residue": str(reward - indexed),
                "userEarned": str(earned),
            }
        )

    redemption_quotes = []
    for case in scenarios["redemptionCases"]:
        shares = int(case["shares"])
        supply = int(case["supplyBefore"])
        redemption_quotes.append(
            {
                "id": case["id"],
                "percentageWad": str(mul_div(shares, WAD, supply)),
                "outputs": [
                    {"asset": asset["asset"], "amount": str(mul_div(int(asset["balance"]), shares, supply))}
                    for asset in case["assets"]
                ],
            }
        )

    supply_quotes = []
    for case in scenarios["supplyCases"]:
        supply = int(case["lifetimeMinted"]) - int(case["lifetimeBurned"])
        change = int(case["gbxMined"]) - int(case["gbxBurned"])
        supply_quotes.append(
            {
                "id": case["id"],
                "currentSupply": str(supply),
                "netSupplyChange": str(change),
                "projectedSupply": str(supply + change),
            }
        )

    return {
        "schemaVersion": scenarios["schemaVersion"],
        "usdGDecimals": scenarios["usdGDecimals"],
        "targetDecimals": scenarios["targetDecimals"],
        "infiniteSupply": True,
        "genesisLiquidityAllocation": str(GENESIS_LP_GBX),
        "miningQuotes": mining_quotes,
        "auctionQuotes": auction_quotes,
        "rewardQuotes": reward_quotes,
        "redemptionQuotes": redemption_quotes,
        "supplyQuotes": supply_quotes,
    }


if __name__ == "__main__":
    path = Path(sys.argv[1])
    print(json.dumps(compute(json.loads(path.read_text())), indent=2))
